### ISSUE TITLE
How does injected.isAuthorized() work?

### ISSUE THREAD

**Issue by sebastiandeutsch** [NoahZinsmeister/web3-react#78]
Hi,

I assume `injected.isAuthorized()` is returning a promise if (e.g. MetaMask has already allowed the connection) if the user is already authorised.

If the user is `isAuthorized` do I have to activate the connector again? If I do so I'll get a 

```
Warning: Suppressed stale connector activation [object Object]
```

Warning.

Thanks for your help

Sebastian

**Comment by sebastiandeutsch** [comment]
More broadly a little more info about the mechanisms would be lovely:

https://github.com/NoahZinsmeister/web3-react/blob/v6/example/pages/index.tsx#L240 

I guess this code

```
  // handle logic to eagerly connect to the injected ethereum provider, if it exists and has granted access already
  const triedEager = useEagerConnect()
```

is for a an injected ethereum provider (that is already authorized) e.g. MetaMask is installed, I'm logged in and have accepted the connection.

```
  // handle logic to connect in reaction to certain events on the injected ethereum provider, if it exists
  useInactiveListener(!triedEager || !!activatingConnector)
```

What does the suppress argument do? Why is this needed at all?

**Comment by hbarcelos** [comment]
The issue I'm getting here is that `isAuthorized()` always returns `false` if I refresh the page, even if I have already given permission.

**Comment by jmahhh** [comment]
Same issue as @hbarcelos

**Comment by jmahhh** [comment]
I dug a little deeper, in 
`injected-connector.cjs.development.js` >`_catch()`
there is an error ` TypeError: "window.ethereum.send(...) is undefined"` which results in `false` for `useEagerConnect`. 

which is kind of weird because if I log `window.ethereum.send` right before it a function is returned. 

The uniswap repo, which apparently gets `true` on `useEagerConnect`, uses `injected-connector.esm.js` and there are no library logs so it's a little hard to proceed.

**Comment by maxaleks** [comment]
Same issue as @hbarcelos and @jmahhh

**Comment by jacobedawson** [comment]
Same issue as @hbarcelos and @jmahhh and @maxaleks

**Comment by mtsalenc** [comment]
I don't seem to be affected by this, isAuthorized returns true for injected provider.

**Comment by rpedroni** [comment]
Still seeing this using same or newer versions as @eccentricexit 
Can confirm the connector is active and authorized but will get the `stale` message nonetheless

**Comment by rpedroni** [comment]
The solution was putting the activation with a timeout. There are some race conditions for loading that happen here so using `setTimeout(() => activate(...), 500)` solved it for me - can only imagine doing this on a window loading event would result the same.

**Comment by NoahZinsmeister** [comment]
hi! library author here, sorry for the (extremely) slow responses; for an update on the state of this library, please see https://github.com/NoahZinsmeister/web3-react/issues/163#issuecomment-1010210571

TLDR: v6 is deprecated and going into maintenance mode. i expect not to make any non-security related changes going forward (however, you're always welcome to! the beauty of open source)

i'll be closing a lot of issues with this message right now, but please feel free to re-open if you think the topic is still relevant!

thanks for using web3-react <3

**Comment by sebastiandeutsch** [comment]
Thanks for the heads up <3

**Comment by ackvf** [comment]
I would still like to know why the setTimeout seemingly fixes it? Would adding some awaits and promises help make it more deterministic? What is the right delay to set? 😁

**Comment by ackvf** [comment]
So, I've got an update

First we were using this

```
if (window?.ethereum?._state?.accounts?.length) {
  await sleepResolve(500)
  await activateConnector()
}
```

But then we found that when this code is deployed, the object `window.ethereum` would be empty immediately after a page refresh.

What's more interesting is that in a browser with disabled cache or DevTools open, or when we performed hard refresh instead, the object would NOT be empty! 🤯 I know.

So, this led me to realise that the problem itself is not actually with the `activate()` function, but there is an actual racing condition within this `window.ethereum._state.accounts` object!

> ### _The problem is not with the `activate()` function, but `window.ethereum._state.accounts` object!_

Basically, _something_ tries to write in this object asynchronously, but our `activate()` function is usualy run before it happens.

After we realised this, we changed the flat sleep into polling mechanism.

```
  await poll(() => window?.ethereum?._state?.accounts?.length !== undefined, 50, 500)
  await activateConnector(localConnector)
```

The poll function is basically just a `while` function that checks the condition and counts down milliseconds.

```
while (!condition()) {
  if (timeLeft <= 0 ) return
  timeLeft -= interval
  await sleep(interval)
}
```

Here is a slightly improved version https://gist.github.com/ackvf/68f992660a5eda645c4671d3599b2acf#file-poll-ts
